File: frontend/src/services/ideaService.ts
import api, { ApiResponse } from './api';
import { Idea, CreateIdeaRequest, UpdateIdeaRequest, IdeasByRound } from '../types';

export const ideaService = {
  async submit(data: CreateIdeaRequest): Promise<Idea> {
    const response = await api.post<ApiResponse<Idea>>('/ideas', data);
    if (!response.data.data) {
      throw new Error(response.data.message || 'Failed to submit idea');
    }
    return response.data.data;
  },

  async update(id: string, data: UpdateIdeaRequest): Promise<Idea> {
    const response = await api.put<ApiResponse<Idea>>(`/ideas/${id}`, data);
    if (!response.data.data) {
      throw new Error(response.data.message || 'Failed to update idea');
    }
    return response.data.data;
  },

  async delete(id: string): Promise<void> {
    await api.delete(`/ideas/${id}`);
  },

  async getByRoundId(roundId: string): Promise<Idea[]> {
    const response = await api.get<ApiResponse<Idea[]>>(`/ideas/round/${roundId}`);
    return response.data.data || [];
  },

  async getBySessionId(sessionId: string): Promise<Idea[]> {
    const response = await api.get<ApiResponse<Idea[]>>(`/ideas/session/${sessionId}`);
    return response.data.data || [];
  },

  async getGroupedByRound(sessionId: string): Promise<IdeasByRound[]> {
    const response = await api.get<ApiResponse<IdeasByRound[]>>(`/ideas/session/${sessionId}/grouped`);
    return response.data.data || [];
  },

  async getMyIdeasInRound(roundId: string): Promise<Idea[]> {
    const response = await api.get<ApiResponse<Idea[]>>(`/ideas/round/${roundId}/my-ideas`);
    return response.data.data || [];
  },

  async canSubmit(roundId: string): Promise<{ canSubmit: boolean; currentCount: number; maxAllowed: number }> {
    const response = await api.get(`/ideas/round/${roundId}/can-submit`);
    return response.data;
  },

  async canSubmitToSession(sessionId: string): Promise<{ canSubmit: boolean }> {
    const response = await api.get(`/ideas/session/${sessionId}/can-submit`);
    return response.data;
  },
};

File: mobile/lib/services/idea_service.dart
import '../core/constants/api_constants.dart';
import '../models/models.dart';
import 'api_client.dart';

class IdeaService {
  final ApiClient _apiClient = ApiClient();

  Future<List<Idea>> getIdeas({String? sessionId, String? roundId}) async {
    final queryParams = <String, dynamic>{};
    if (sessionId != null) queryParams['sessionId'] = sessionId;
    if (roundId != null) queryParams['roundId'] = roundId;

    final response = await _apiClient.get(
      ApiConstants.ideas,
      queryParameters: queryParams.isNotEmpty ? queryParams : null,
    );
    final List<dynamic> data = response.data;
    return data.map((json) => Idea.fromJson(json)).toList();
  }

  Future<Idea> getIdea(String id) async {
    final response = await _apiClient.get('${ApiConstants.ideas}/$id');
    return Idea.fromJson(response.data);
  }

  Future<Idea> createIdea(CreateIdeaRequest request) async {
    final response = await _apiClient.post(
      ApiConstants.ideas,
      data: request.toJson(),
    );
    return Idea.fromJson(response.data);
  }

  Future<Idea> updateIdea(String id, UpdateIdeaRequest request) async {
    final response = await _apiClient.put(
      '${ApiConstants.ideas}/$id',
      data: request.toJson(),
    );
    return Idea.fromJson(response.data);
  }

  Future<void> deleteIdea(String id) async {
    await _apiClient.delete('${ApiConstants.ideas}/$id');
  }

  // Get ideas by round for a session
  Future<List<IdeasByRound>> getIdeasByRound(String sessionId) async {
    final response = await _apiClient.get('${ApiConstants.ideas}/session/$sessionId/by-round');
    final List<dynamic> data = response.data;
    return data.map((json) => IdeasByRound.fromJson(json)).toList();
  }

  // Get my ideas for a session
  Future<List<Idea>> getMyIdeas(String sessionId) async {
    final response = await _apiClient.get('${ApiConstants.ideas}/session/$sessionId/my');
    final List<dynamic> data = response.data;
    return data.map((json) => Idea.fromJson(json)).toList();
  }
}
